README.rst


``webdrivermanager``
=======================

Python module to facilitate downloading and deploying `WebDriver <https://www.w3.org/TR/webdriver/>`_ binaries.  The classes in this module can be used to automatically search for and download the latest version (or a specific version) of a WebDriver binary and then extract it and place it by copying or symlinking it to the location where Selenium or other tools should be able to find it then.


Installation
------------

This module is available on the Python Package Index (PyPI) and can be installed as follows:

``pip install webdrivermanager``


Dependencies
------------

This module is dependent on the following additional packages:

- `requests <https://pypi.org/project/requests/>`_
- `tqdm <https://pypi.org/project/tqdm/>`_
- `BeautifulSoup4 <https://pypi.org/project/beautifulsoup4/>`_
- `appdirs <https://pypi.org/project/appdirs/>`_


Classes
-------

The following classes are available:

- ``ChromeDriverManager`` for downloading and installing `chromedriver <https://sites.google.com/a/chromium.org/chromedriver/downloads>`_ (for Google Chrome).
- ``GeckoDriverManager`` for downloading and installing `geckodriver <https://github.com/mozilla/geckodriver>`_ (for Mozilla Firefox).
- ``OperaChromiumDriverManager`` for downloading and installing `operadriver <https://github.com/operasoftware/operachromiumdriver>`_ (for Chromium based Opera browsers).
- ``EdgeDriverManager`` for downloading and installing `edgedriver <https://developer.microsoft.com/en-us/microsoft-edge/tools/webdriver/>`_ (for Microsoft Edge).
- ``EdgeChromiumDriverManager`` for downloading and installing Edge Chromium based webdrivers
- ``IeDriverManager`` for downloading and installing Internet Explorer based webdrivers


Status
------

Currently being developed/tested using Python 2.7.15 and 3.7  on macOS, Windows & Linux


Example module usage
--------------------

Example::

   >>> from webdrivermanager import GeckoDriverManager
   >>> gdd = GeckoDriverManager()
   >>> gdd.download_and_install()
   1524kb [00:00, 1631.24kb/s]
   ('/Users/rasjani/webdriver/geckodriver-v0.20.1-macos/geckodriver', '/Users/rasjani/bin/geckodriver')
   >>> gdd.download_and_install("v0.20.0")
   1501kb [00:02, 678.92kb/s]
   Symlink /Users/rasjani/bin/geckodriver already exists and will be overwritten.
   ('/Users/rasjani/webdriver/geckodriver-v0.20.0-macos/geckodriver', '/Users/rasjani/bin/geckodriver')
   >>> gdd.download_and_install()
   Symlink /Users/rasjani/bin/geckodriver already exists and will be overwritten.
   ('/Users/rasjani/webdriver/geckodriver-v0.20.1-macos/geckodriver', '/Users/rasjani/bin/geckodriver')
   >>>


Command line tool
-----------------

There is a command-line tool that is also available.  After installing the package, it can be used as follows (Windows example)::

   > webdrivermanager chrome:2.38 firefox opera:v.2.35
   Downloading WebDriver for browser: 'chrome'
   3300kb [00:00, 11216.38kb/s]
   Driver binary downloaded to: C:\Users\rasjani\webdriver\chrome\2.38\2.38%2Fchromedriver_win32\chromedriver.exe
   Driver copied to: C:\Users\rasjani\bin\chromedriver.exe

   Downloading WebDriver for browser: 'firefox'
   3031kb [00:01, 2253.64kb/s]
   Driver binary downloaded to: C:\Users\rasjani\webdriver\gecko\v0.20.1\geckodriver-v0.20.1-win64\geckodriver.exe
   Driver copied to: C:\Users\rasjani\bin\geckodriver.exe

   Downloading WebDriver for browser: 'opera'
   3548kb [00:02, 1239.02kb/s]
   Driver binary downloaded to: C:\Users\rasjani\webdriver\operachromium\v.2.35\operadriver_win64\operadriver_win64\operadriver.exe
   Driver copied to: C:\Users\rasjani\bin\operadriver.exe

   WARNING: Path 'C:\Users\rasjani\bin' is not in the PATH environment variable.

In the above example, a version was specified for Chrome and Opera while no version was specified for Firefox so the latest version of ``geckodriver`` was implicitly downloaded.

Command line options
--------------------

    usage: webdrivermanager [-h] [--downloadpath F] [--linkpath F] [--os OSNAME]
              browser [browser ...]

        Tool for downloading and installing WebDriver binaries.

	positional arguments:
	   browser               Browser to download the corresponding WebDriver
				 binary. Valid values are: chrome, firefox, gecko,
				 mozilla, opera, edge. Optionally specify a version
				 number of the WebDriver binary as follows:
				 'browser:version' e.g. 'chrome:2.39'. If no version
				 number is specified, the latest available version of
				 the WebDriver binary will be downloaded.

        optional arguments:
            -h, --help            show this help message and exit
            --downloadpath F, -d F
                                  Where to download the webdriver binaries
            --linkpath F, -l F    Where to link the webdriver binary to. Set to "AUTO"
                                  if you need some intelligence to decice where to place
                                  the final webdriver binary
            --linkpath F, -l F    Where to link the webdriver binary to. Set to "AUTO"
                                  if you need some intelligense to decide where to place
                                  the final webdriver binary. If set to "SKIP", no
                                  link/copy done
            --os OSNAME, -o OSNAME
                                  Overrides os detection with given os name


Do note that `--downloadpath`/`-d` flag location is used for storing the whole downloaded and then `--linkpath`/`-l` path location is where the final binary is either symlinled or copied to.  Linkpath should be the directory you either already have in PATH or you should place there since tools using these webdrivers usually locate the appropriate webdriver binary from PATH environment variable.

If linkpath flag is set to *AUTO*, tool will iterate over your current PATH environment variable and tries to find the first writeable directory within it and place the copy or symlink into it. If linkpath is set to *SKIP*, only download is done, linking/copying is skipped.

License
-------

This is released under an MIT license.  See the ``LICENSE`` file in this repository for more information.

Consult the license terms of the providers of the WebDriver downloads prior to downloading / using the WebDrivers.


CHANGELOG
* Mon Feb 22 2021 Jani Mikkonen <jani.mikkonen@gmail.com> - 0.10.0
- Add baseclass methods for parsing latest and compatible versions (rev.9d37e559)
- Code restructure and refactorings (rev.ded7c464)
- Add option for 'compatible' chromedriver version (rev.537edf64)
- Version detection/download on EdgeChromium fixed. (rev.72c9955f)
- Remove 2.7 support from CI (rev.c612eefe)

* Sun Sep 06 2020 Jani Mikkonen <jani.mikkonen@gmail.com> - 0.9.0
- fix linux geckodriver download (rev.778b1d1c)
- Fix support for EdgeChromium (rev.77485c05)
- - Fix support for IE --bitness 32 (rev.288d96da)

* Sat Mar 14 2020 Jani Mikkonen <jani.mikkonen@gmail.com> - 0.8.0
- Release 0.8.0 (rev.758af05f)
- Support for "EdgeChromium" webdrivers. (rev.5f8f8e5d)
- Fix dev dependencies to support 2.7 python still. (rev.14b37f45)
- dev deps updated (rev.079a3004)
- Added support for tox (rev.613cdf37)
- Use fixed gitchangelog from own fork (rev.a6a174ee)
- Few lintian fixes (rev.5ac893f2)
- Bugfix: AVAILABLE_DRIVERS exported incorrectly. (rev.17bf7453)
- Basic code for Internet Explorer (rev.d069e239)
- Fixes new edge driver download scheme .. (rev.6670b251)
- include licese file in package (rev.961060ad)
- expose get_version (rev.80e43249)

* Thu Mar 14 2019 Jani Mikkonen <jani.mikkonen@siili.com> - 0.7.4
- Version bump to 0.7.4 (rev.181143ff)
- Fixed imports (rev.50f96502)
- Use flaky package go retry tests failed (rev.8d38107b)
- Use better reporters for pylint&flake8 (rev.e4bfedc0)

* Wed Mar 13 2019 Jani Mikkonen <jani.mikkonen@siili.com> - 0.7.3
- Version bump to 0.7.3 (rev.c8a2926a)
- Show human readable message on connection error (rev.dc856570)
- Changelog into a separate file (rev.76b3fbf6)
- Add version information (rev.c9074259)
- lxml removed from pylintrc (rev.6553c556)
- Fixed readme (rev.d46c8453)

* Wed Feb 20 2019 Jani Mikkonen <jani.mikkonen@siili.com> - 0.7.2
- Version bump to 0.7.2 (rev.c738289e)
- Bugfix: unhandled exception if windows binary did not exists (rev.c23720f3)

* Tue Feb 19 2019 Jani Mikkonen <jani.mikkonen@siili.com> - 0.7.1
- version bump (rev.852a3651)
- Cygwin bug fixes (rev.a3b3297b)
- Add option to skip link creation (rev.54c4b309)
- Enable choosing binary bitness (rev.0ac1a7e4)
- moar lintian fixes (rev.aa40df9c)
- Bugfix: handle chromedriver bitness across platforms (rev.07940b7d)
- Fix for python2/3 compatibility (rev.9a4c020e)
- Fix for requirements (rev.884727ce)
- Rewrite for get_download_url (rev.9952ec63)
- Updated requirements-dev (rev.0d70fdae)
- Updated readme to include edge (rev.f005d076)
- Pylint fixes (rev.1eed6bce)
- BeautifulSoup instead of lxml (rev.b0fcdc24)
- Run flake8 & pylint on ci and generate junits (rev.213136ae)
- Cygwin support (rev.367d9782)
- Extended tests and lots of flake8 fixes (rev.9fa5faeb)
- updated readme (rev.5b7e3b49)
- Overwrite also binary if it exists (rev.528f703d)
- option to autodiscover bin location (rev.8ea9817f)

* Tue Jan 29 2019 Jani Mikkonen <jani.mikkonen@siili.com> - 0.6.0
- Refactor cli portion (rev.ee43c164)
- Refactor cli portion (rev.dbf95149)

* Fri Jan 04 2019 Jani Mikkonen <jani.mikkonen@siili.com> - 0.5.0
- Export test results to azure (rev.f5fbfc45)
- Version bump to 0.5.0 (rev.00f66be2)
- Dont crash with unknown browser string (rev.8118dab2)
- Basic support for edge (rev.e9af3942)
- Add possibility to download any os binaries (rev.4a705c0a)
- Removed dead code (rev.dbd166d6)

* Fri Nov 02 2018 Jani Mikkonen <jani.mikkonen@siili.com> - 0.4.2
- Release: 0.4.2 (rev.d01a3c35)
- Corner case for Mac OS X (rev.e4210de7)
- Dont publish artifacts (rev.c3233343)
- Fixed fallback github method (rev.f712f36d)

* Wed Oct 31 2018 Jani Mikkonen <jani.mikkonen@siili.com> - 0.4.1
- version bump (rev.7a2c4f0c)
- Fix & release (rev.d9cdaf2e)
- code deduplication (rev.aa5417aa)
- doit tasks and few dependencies for dev side (rev.37a250b7)
- publish artifacts task (rev.79d324ae)
- python 2.7 fix (rev.fc7f4e9e)
- Removed cruft (rev.d60897eb)
- Basic pipeline (rev.eba22e39)
- Fix download_and_install show_progress_bar handling (rev.931a2457)
- acceptance tests to just prevent things getting broken (rev.e3907313)
- Bug fixes: (rev.7f560bb8)
- All Downloaders into Managers (rev.b9f3c0f0)

* Sun Oct 28 2018 Jani Mikkonen <jani.mikkonen@siili.com> - 0.4.0
- Version bump to 0.4.0 (rev.1db2cb4d)
- Fixed chrome/opera download in mac (rev.cdd21b92)
- setup.py pulls requirements.txt (rev.70be9c34)
- Added lxml dependency (rev.78cee05d)
- Added a fallback download method for opera chromium driver (rev.03766304)
- Added a fallback download method for geckodriver (rev.f4a364ea)

* Sat Oct 27 2018 Jani Mikkonen <jani.mikkonen@siili.com> - 0.3.0
- Version bump (rev.e368c12f)
- Updated readme (rev.ee237b21)
- cache get_download_path calls (rev.3749cc01)
- Updated readme (rev.09c3569f)

* Fri Oct 26 2018 Jani Mikkonen <jani.mikkonen@siili.com> - 0.2.0
- utf8 for the win (rev.26d907aa)
- support passing dl and linkpath from cli (rev.2a89467c)
- Misc changes (rev.fea92ae7)
- Setup changes (rev.3dc12714)
- dev requirements (rev.31015de3)
- Fix setup.py (rev.a3bdf592)
- If venv is activate, choose install location accordingly. (rev.66d91386)
- Rename the poroject (rev.86cd02b5)
- Fix up dependencies, add a few more classifiers to setup. (rev.88761b17)
- Updated README.rst to include info about command line tool, downloaders for Chrome and Chromium-based Opera browsers. (rev.d9779a68)
- Finish adding support for downloading a specific WebDriver version to the command-line tool. (rev.b1a2a6ae)
- Fix display bug if link_path is not in the PATH environment variable. (rev.fdd785e0)
- Add command-line support. (rev.b2a9d735)
- Initial work on OperaChromiumDriverDownloader class. Change download behavior to include webdriver name and version in the download path (needed since the version number is not embedded in the filename for the Opera driver). Fix bug where the executable bit was not getting set on the extracted driver binary. (rev.8f6c34ff)

webdrivermanager/cli.py
# -*- coding: utf-8 -*-

import os
import sys
import os.path
import argparse
from requests import ConnectionError

try:
    from ._version import get_versions
except ImportError:
    from _version import get_versions
try:
    from .webdrivermanager import AVAILABLE_DRIVERS as DOWNLOADERS
except ImportError:
    from webdrivermanager import AVAILABLE_DRIVERS as DOWNLOADERS


OS_NAMES = ['mac', 'win', 'linux']
BITNESS = ["32", "64"]


def parse_command_line():
    parser = argparse.ArgumentParser(
        description='Tool for downloading and installing WebDriver binaries. Version: {}'.format(get_versions()["version"]),
    )
    parser.add_argument('browser', help='Browser to download the corresponding WebDriver binary.  Valid values are: {0}. Optionally specify a version number of the WebDriver binary as follows: \'browser:version\' e.g. \'chrome:2.39\'.  If no version number is specified, the latest available version of the WebDriver binary will be downloaded.'.format(', '.join(DOWNLOADERS.keys())), nargs='+')
    parser.add_argument('--downloadpath', '-d', action='store', dest='downloadpath', metavar='F', default=None, help='Where to download the webdriver binaries')
    parser.add_argument('--linkpath', '-l', action='store', dest='linkpath', metavar='F', default=None, help='Where to link the webdriver binary to. Set to "AUTO" if you need some intelligense to decide where to place the final webdriver binary. If set to "SKIP", no link/copy done.')
    parser.add_argument('--os', '-o', action='store', dest='os_name', choices=OS_NAMES, metavar='OSNAME', default=None, help='Overrides os detection with given os name. Values: {0}'.format(', '.join(OS_NAMES)))
    parser.add_argument('--bitness', '-b', action='store', dest='bitness', choices=BITNESS, metavar='BITS', default=None, help='Overrides bitness detection with given value. Values: {0}'.format(', '.join(BITNESS)))
    parser.add_argument('--version', action='version', version='%(prog)s {}'.format(get_versions()["version"]))
    return parser.parse_args()


def main():
    args = parse_command_line()
    for browser in args.browser:
        if ':' in browser:
            browser, version = browser.split(':')
        else:
            version = 'latest'
        if browser.lower() in DOWNLOADERS.keys():
            print('Downloading WebDriver for browser: "{0}"'.format(browser))
            downloader = DOWNLOADERS[browser](args.downloadpath, args.linkpath, args.os_name, args.bitness)
            try:
                extracted_binary, link = downloader.download_and_install(version)
            except ConnectionError:
                print("Unable to download webdriver's at this time due to network connectivity error")
                sys.exit(1)
            print('Driver binary downloaded to: "{0}"'.format(extracted_binary))
            if link is not None:
                if os.path.islink(link):
                    print('Symlink created: {0}'.format(link))
                else:
                    print('Driver copied to: {0}'.format(link))
                link_path = os.path.split(link)[0]
                if link_path not in os.environ['PATH'].split(os.pathsep):
                    print('WARNING: Path "{0}" is not in the PATH environment variable.'.format(link_path))
            else:
                print('Linking webdriver skipped')
        else:
            print('Unrecognized browser: "{0}".  Ignoring...'.format(browser))
        print('')


if __name__ == '__main__':
    main()
